staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cel
+
+import (
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/checker/decls"
+	"github.com/google/cel-go/ext"
+
+	expr "google.golang.org/genproto/googleapis/api/expr/v1alpha1"
+	"google.golang.org/protobuf/proto"
+
+	"k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	celmodel "k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model"
+)
+
+// ScopedTypeName is the placeholder type name used for the type of ScopedVarName if it is an object type.
+const ScopedTypeName = "apiextensions.k8s.io.v1alpha1.ValidationRuleSelf"
+
+// ScopedVarName is the variable name assigned to the locally scoped data element of a CEL valid.
+const ScopedVarName = "self"
+
+// CompilationResult represents the cel compilation result for one rule
+type CompilationResult struct {
+	Program cel.Program
+	Error   *Error
+}
+
+// Compile compiles all the XValidations rules (without recursing into the schema) and returns a slice containing a compiled program for each provided CelRule, or an array of errors.
+func Compile(s *schema.Structural, isResourceRoot bool) ([]CompilationResult, error) {
+	if len(s.Extensions.XValidations) == 0 {
+		return nil, nil
+	}
+	celRules := s.Extensions.XValidations
+
+	var propDecls []*expr.Decl
+	var root *celmodel.DeclType
+	var ok bool
+	env, err := cel.NewEnv()
+	if err != nil {
+		return nil, err
+	}
+	reg := celmodel.NewRegistry(env)
+	rt, err := celmodel.NewRuleTypes(ScopedTypeName, s, isResourceRoot, reg)
+	if err != nil {
+		return nil, err
+	}
+	if rt == nil {
+		return nil, nil
+	}
+	opts, err := rt.EnvOptions(env.TypeProvider())
+	if err != nil {
+		return nil, err
+	}
+	root, ok = rt.FindDeclType(ScopedTypeName)
+	if !ok {
+		root = celmodel.SchemaDeclType(s, isResourceRoot).MaybeAssignTypeName(ScopedTypeName)
+	}
+	propDecls = append(propDecls, decls.NewVar(ScopedVarName, root.ExprType()))
+	opts = append(opts, cel.Declarations(propDecls...))
+	opts = append(opts, ext.Strings())
+	env, err = env.Extend(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	// compResults is the return value which saves a list of compilation results in the same order as x-kubernetes-validations rules.
+	compResults := make([]CompilationResult, len(celRules))
+	for i, rule := range celRules {
+		var compilationResult CompilationResult
+		if rule.Rule == "" {
```

Comment:
make sure the rule actually has content
```suggestion
		if len(strings.TrimSpace(rule.Rule)) == 0 {
```